### DISCUSSION TITLE
iPXE boot from USB on Apple Mac Mini (2012, Intel)

### DISCUSSION THREAD

**Discussion by bplein** [ipxe/ipxe#1269] [Q&A]
Hello! 

I am trying to use iPXE to help boot a few old 2012 Mac mini servers into Linux. Using the generic ipxe.usb -> USB, I can manually choose the boot option for the device, and iPXE works as advertised. 

2012 Mac minis won't automatically choose and boot from that USB device even if there are no other bootable devices. If I recall correctly, it's because that bootable USB is not GPT.

I have also tried using the `bin-x86_64-efi/ipxe.efirom` method, then copying that to an appropriately formatted USB drive as `bootx64.efi`, which IS automatically bootable, and starts. But it hangs when getting a DHCP address, and if I replace `autoexec.ipxe` with a script that hops to shell, the boot process hops to an iPXE shell as desired, but no keyboard input.

My guess is that this has something to do with lack of appropriate USB drivers (for the keyboard) in the`ipxe.efirom` file, and general strangeness with Apple EFI implementation. 

Has anyone worked around this? I spent hours last night building debug versions, trying to convert the `.usb` version into a GPT based disk (not EFI) as it seems to be working in BIOS mode, etc.  I'd be fine with BIOS mode boot but it needs to NOT be an MBR formatted disk.

**Comment by NiKiZe** [comment]
Default build of iPXE does not have USB at all, instead firmware is expected to provide input and output thru console.

If this is a GPT issue, you should be able to add that to the .usb build, you might need to resize the image to create space at the end of disk to hold the GPT partition data.
You might be able to remove MBR partition data, but I would expect that to not matter (GPT + MBR is valid and the most common setup) however, do remember that with MBR boot, the first stage is actually executing the MBR, which then usually chain to elsewhere on disk,  partitions isn't actually a requirement.

I would recommend trying snp.efi, or snponly.efi as bootx64.efi, as ipxe has native drivers for broadcom, which I know isn't working great. (Open issue)

**Comment by bplein** [comment]
Thanks, I appreciate the input. I just tried using snp.efi as `bootx64.efi`, along with a simple autoexec.ipxe: 
```
#!ipxe
echo Booting from EFI ...
shell
```
When booting from this USB drive, it picks up the autoexec.ipxe, prints "Booting from EFI ...", and drops to shell, but my USB keyboard is still unusable. 

This sounds very similar to the other discussion about a Cisco UCS failing in the same way.  

So am I correct in saying that the reason booting iPXE from BIOS works, but booting via EFI does not, is that the BIOs emulation includes keyboard support while Apple's EFI implementation requires addtional EFI support for keyboard?

**Reply by NiKiZe** [reply]
I'd guess that Apples efi firmware don't handle or don't expose input. While for pcbios emulation it is forced to.

You could test ecm.efi or any other build that includes USB driver, which then also should enable the native iPXE USB HID stack. One likely probability tho, is that the firmware won't allow iPXE to take over USB at all. (We're talking about Apple here...)

**Reply by bplein** [reply]
Yeah, I don't know enough about the low level details. I just wish the ipxe.usb was a bootable GUID/GPT format.  I don't need EFI, and the non-EFI pixel works fine except for not being bootable automatically. 

And while I'm a wiz with dd, I'm not a wizard with how to transfer the contents of an MBR partition to GPT.  Sure I can dd it and have tried, but that's not sufficient.

**Reply by NiKiZe** [reply]
It should however be sufficient to run `fdisk ipxe.usb` and choose something like "save GPT partition table"

This might violate EFI spec...

**Reply by bplein** [reply]
if it works for my lab, I'm good with it. I'm just trying to boot iPXE via USB, so I can allow my tftpboot server to serve up whatever I want (current goals, running Flatcar Linux from memory, not from disk). So it just needs to do the job and get out of the way, only on reboot. Thanks for the ideas, I'll try them later today.

**Reply by marr** [reply]
@bplein i'm curious did you ever get this working? I am trying to install talos on the same mac mini and having trouble
